src/include/miopen/backend_api.hpp

```diff
@@ -0,0 +1,62 @@
+#pragma once
+
+#include <miopen/miopen.h>
+#include <vector>
+
+namespace miopen {
+
+namespace backend_api {
+
+class BackendDescriptor
+{
+public:
+    BackendDescriptor() : m_IsFinalized(false) {}
+    virtual ~BackendDescriptor()                               = default;
+    virtual miopenStatus_t SetAttribute(miopenBackendAttributeName_t attributeName,
+                                        miopenBackendAttributeType_t attributeType,
+                                        int64_t elementCount,
+                                        void* arrayOfElements) = 0;
+    virtual miopenStatus_t Finalize()                          = 0;
+
+protected:
+    bool m_IsFinalized;
+};
+
+class BackendTensorDescriptor : public BackendDescriptor
+{
+public:
+    BackendTensorDescriptor();
+    virtual ~BackendTensorDescriptor() override;
+    virtual miopenStatus_t SetAttribute(miopenBackendAttributeName_t attributeName,
+                                        miopenBackendAttributeType_t attributeType,
+                                        int64_t elementCount,
+                                        void* arrayOfElements) override;
+    virtual miopenStatus_t Finalize() override;
+    inline miopenStatus_t Status() const noexcept { return m_Status; }
+    inline bool IsVirtual() const noexcept { return m_IsVirtual; }
+
+private:
+    miopenTensorDescriptor_t m_Descriptor;
+    std::vector<int64_t> m_Dimensions;
```

Comment:
We are duplicating information here that will eventually be in `miopen::TensorDescriptor`. I can see why `Finalize()` requires doing so but let's think of ways to get rid of it.